## 2.4 Metrics - nDCG Exercise

Objective of the exercise is to compute the nDCG for two rankings of those documents and determine which is best.

First we need some data - the file 'data/query-for-ndcg.csv' contains a set of documents and relevance score for a given query. Here we load the data:

In [129]:
import pandas as pd

df = pd.read_csv('data/query-for-ndcg.csv')
df.head()

score                                              title
0    3.2  new akslen modle flashing safety led bike bicy...
1    4.4  usb led bike bicycle cycling warning tail rear...
2    1.5  flash led neon bicycle wheels tire valve dust ...
3    4.8  black cycling bike bicycle led front head ligh...
4    4.2  cycling bicycle taillights led safety flashing...

Two different rankers return the documents in two different orders.  We can compute the relevance score vector for each ranker: 

In [130]:
r1 = [ 19, 3, 1, 10, 18, 0, 15, 14, 7, 17, 16, 6, 4, 13,12, 8, 2, 9, 5 ]
r2 = [ 19, 3, 1, 10, 18, 0, 8 , 14, 7, 17, 16, 6, 4, 13,12, 15, 2, 9, 5 ]

scores1 = df.ix[r1,'score'].values
scores2 = df.ix[r2,'score'].values

## Computing nDCG

<img src='files/resources/ic_assignment_black_24dp_2x.png' align='left'>Write a function to compute nDCG.  The function should take a single parameter which is the list of relevance scores.  
Compute nDCG for the csore vectors scores1 and scores2.

If the items returned by the ranker had relevance scores (5.0,4.1,4.0,2.3,1.0) we should be able to call the ndcg() function in in this manner `ndcg([5.0, 4.1, 4.0, 2.3, 1.0])`.


In [126]:
import math

def dcg(scores):
    dcg = 0
    for i,s in enumerate(scores):
        dcg += scores[i] / math.log(i + 2,2)
    return dcg

def ndcg(scores):
    return dcg(scores) / dcg(sorted(scores, reverse=True))


0.966896361402
0.980459274646


Compute nDCG for R1 and R2:

In [128]:
ndcg_r1 = ndcg(scores1)
ndcg_r2 = ndcg(scores2)

print('nDCG R1 = {0:6.2f}'.format(ndcg_r1))
print('nDCG R1 = {0:6.2f}'.format(ndcg_r2))

nDCG R1 =   0.97
nDCG R1 =   0.98


If everything worked then you should have got an nDCG score of 0.97 for R1 and 0.98 for R2.  